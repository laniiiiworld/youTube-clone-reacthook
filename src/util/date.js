import { format, register } from 'timeago.js';
import koLocale from 'timeago.js/lib/lang/ko';

register('ko-KR', koLocale);

export function formatAgo(date) {
  return format(date, navigator.language === 'ko-KR' ? 'ko-KR' : 'en-US');
}
