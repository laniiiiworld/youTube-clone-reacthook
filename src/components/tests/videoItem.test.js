import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route, useLocation } from 'react-router-dom';
import { withRouter } from '../../tests/utils';
import { fakeVideo as video } from '../../tests/videos';
import { formatAgo } from '../../util/date';
import VideoItem from '../videoItem/videoItem';

describe('VideoItem', () => {
  const { title, channelTitle, publishedAt, thumbnails } = video.snippet;

  it('renders grid type correctly', () => {
    const component = renderer.create(withRouter(<Route path='/' element={<VideoItem video={video} />} />));
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders list type correctly', () => {
    const component = renderer.create(withRouter(<Route path='/' element={<VideoItem video={video} displayType='list' />} />));
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders video items', () => {
    render(withRouter(<Route path='/' element={<VideoItem video={video} />} />));

    const image = screen.getByRole('img');
    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(screen.getByText(formatAgo(publishedAt))).toBeInTheDocument();
  });

  it('navigates to detailed video page with video state when clicked', async () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }

    render(
      withRouter(
        <>
          <Route path='/' element={<VideoItem video={video} />} />
          <Route path={`/detail/${video.id}`} element={<LocationStateDisplay />} />
        </>
      )
    );

    const item = screen.getByRole('listitem');
    await waitFor(() => userEvent.click(item));

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
