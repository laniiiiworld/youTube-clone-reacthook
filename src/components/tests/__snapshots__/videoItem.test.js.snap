// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoItem renders grid type correctly 1`] = `
<li
  className="videoItem grid"
  data-testid="video-list"
  onClick={[Function]}
>
  <div
    className="thumbnail"
  >
    <img
      alt="title"
      className="thumbnailImg"
      src="http://image/"
    />
  </div>
  <div
    className="info"
  >
    <p
      className="title"
    >
      title
    </p>
    <p
      className="channelTitle"
    >
      channelTitle
    </p>
    <p
      className="publishedAt"
    >
      just now
    </p>
  </div>
</li>
`;

exports[`VideoItem renders list type correctly 1`] = `
<li
  className="videoItem list"
  data-testid="video-list"
  onClick={[Function]}
>
  <div
    className="thumbnail"
  >
    <img
      alt="title"
      className="thumbnailImg"
      src="http://image/"
    />
  </div>
  <div
    className="info"
  >
    <p
      className="title"
    >
      title
    </p>
    <p
      className="channelTitle"
    >
      channelTitle
    </p>
    <p
      className="publishedAt"
    >
      just now
    </p>
  </div>
</li>
`;
